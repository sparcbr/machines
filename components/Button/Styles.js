import { StyleSheet, Platform } from 'react-native'
import { globals } from '../../assets/styles/Globals'
import { colors } from '../../assets/styles/Colors'

const height = 55
const miniHeight = 30

export const styles = StyleSheet.create({
	button: {
		justifyContent: 'center',
		paddingLeft: globals.bigButtons.lateralPadding,
		paddingRight: globals.bigButtons.lateralPadding,
		height: height,
		borderRadius: globals.bigButtons.borderRadius,
		textAlign: 'center'
	},
	miniButton: {
		height: miniHeight,
		justifyContent: 'center',
		paddingLeft: 15,
		paddingRight: 15,
		borderRadius: 20,
		textAlign: 'center'
	},
	defaultButton: {
		backgroundColor: colors.white
	},
	defaultButtonText: {
		color: globals.bigButtons.color,
		fontFamily: globals.bigButtons.fontFamily,
		fontSize: globals.bigButtons.fontSize,
		lineHeight: globals.bigButtons.lineHeight,
		letterSpacing: globals.bigButtons.letterSpacing,
		textTransform: globals.bigButtons.textTransform,
		textAlign: 'center'
	},
	miniLightWhiteButton: {
		borderColor: colors.white,
		borderWidth: 1
	},
	miniButtonText: {
		color: colors.white,
		fontFamily: globals.smallButtons.fontFamily,
		fontSize: globals.smallButtons.fontSize,
		lineHeight: globals.smallButtons.lineHeight,
		letterSpacing: globals.smallButtons.letterSpacing,
		textTransform: globals.smallButtons.textTransform
	},
	blueButton: {
		backgroundColor: colors.blue
	},
	blueButtonText: {
		color: colors.white
	},
	yellowButtonText: {
		color: colors.nightBlue
	},
	greenButtonText: {
		color: colors.nightBlue,
		...Platform.select({
			ios: {
				paddingTop: 3
			}
		})
	}
})
