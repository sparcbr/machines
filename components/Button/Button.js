import React from 'react'
import { TouchableOpacity, Text, View } from 'react-native'
import LinearGradient from 'react-native-linear-gradient'
import { colors } from '../../assets/styles/Colors'
import { Spinner } from '../Spinner/Spinner'
import { styles } from './Styles'

function Button(props) {
	const {
		onPress,
		isLoading,
		label,
		children,
		styleData,
		spinnerColor,
		theme
	} = props
	const buttonStyles = []
	function renderYellowTheme() {
		return (
			renderLoader() || (
				<LinearGradient
					colors={[colors.orange, colors.yellow]}
					start={{ x: 0, y: 0 }}
					end={{ x: 1, y: 0 }}
					style={styles.button}
				>
					{renderCta(styles.yellowButtonText)}
				</LinearGradient>
			)
		)
	}
	function renderGreenTheme() {
		return (
			renderLoader() || (
				<LinearGradient
					colors={[colors.green, colors.lightGreen]}
					start={{ x: 0, y: 0 }}
					end={{ x: 1, y: 0 }}
					style={styles.button}
				>
					{renderCta(styles.greenButtonText)}
				</LinearGradient>
			)
		)
	}
	function renderBlueTheme() {
		return (
			renderLoader() || (
				<View style={{ ...styles.button, ...styles.blueButton }}>
					{renderCta(styles.blueButtonText)}
				</View>
			)
		)
	}
	function renderDefaultTheme() {
		return (
			renderLoader() || (
				<View style={{ ...styles.button, ...styles.defaultButton }}>
					{renderCta()}
				</View>
			)
		)
	}
	function renderMiniLightWhiteTheme() {
		return (
			renderLoader() || (
				<View style={{ ...styles.miniButton, ...styles.miniLightWhiteButton }}>
					{renderCta(styles.miniButtonText)}
				</View>
			)
		)
	}
	function renderCta(additionalTextStyle) {
		const textStyles = [styles.defaultButtonText, additionalTextStyle]

		if (styleData && styleData.text) {
			textStyles.push(styleData.text)
		}

		return label ? (
			<Text
				numberOfLines={1}
				ellipsizeMode={'tail'}
				adjustsFontSizeToFit={true}
				minimumFontScale={0.7}
				style={textStyles}
			>
				{label}
			</Text>
		) : (
			children
		)
	}
	function renderLoader() {
		if (isLoading) {
			return (
				<View style={styles.button}>
					<Spinner color={spinnerColor} />
				</View>
			)
		}

		return null
	}
	function renderContent() {
		let content

		switch (theme) {
			case 'miniLightWhite':
				content = renderMiniLightWhiteTheme()
				break

			case 'green':
				content = renderGreenTheme()
				break

			case 'yellow':
				content = renderYellowTheme()
				break

			case 'blue':
				content = renderBlueTheme()
				break

			default:
				content = renderDefaultTheme()
				break
		}

		return content
	}

	if (styleData && styleData.button) buttonStyles.push(styleData.button)

	return (
		<TouchableOpacity
			disabled={isLoading}
			onPress={e => onPress(e)}
			style={buttonStyles}
		>
			{renderContent()}
		</TouchableOpacity>
	)
}

export { Button }
