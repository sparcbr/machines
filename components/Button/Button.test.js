import React from 'react'
import renderer from 'react-test-renderer'
import { Button } from './Button'

describe('when <Button /> is mounted', () => {
	it('renders label correctly', () => {
		const tree = renderer.create(<Button label="My CTA" />).toJSON()
		expect(tree).toMatchSnapshot()
	})

	it('renders no label button correctly', () => {
		const tree = renderer.create(<Button />).toJSON()
		expect(tree).toMatchSnapshot()
	})

	it('renders miniLightWhite theme correctly', () => {
		const tree = renderer.create(<Button theme="miniLightWhite" />).toJSON()
		expect(tree).toMatchSnapshot()
	})

	it('renders green theme correctly', () => {
		const tree = renderer.create(<Button theme="green" />).toJSON()
		expect(tree).toMatchSnapshot()
	})

	it('renders yellow theme correctly', () => {
		const tree = renderer.create(<Button theme="yellow" />).toJSON()
		expect(tree).toMatchSnapshot()
	})

	it('renders blue theme correctly', () => {
		const tree = renderer.create(<Button theme="blue" />).toJSON()
		expect(tree).toMatchSnapshot()
	})

	it('renders default theme correctly', () => {
		const tree = renderer.create(<Button />).toJSON()
		expect(tree).toMatchSnapshot()
	})
})
