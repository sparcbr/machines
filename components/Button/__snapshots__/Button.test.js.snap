// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when <Button /> is mounted renders blue theme correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#05b0ff",
        "borderRadius": 35,
        "height": 55,
        "justifyContent": "center",
        "paddingLeft": 30,
        "paddingRight": 30,
        "textAlign": "center",
      }
    }
  />
</View>
`;

exports[`when <Button /> is mounted renders default theme correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#ffffff",
        "borderRadius": 35,
        "height": 55,
        "justifyContent": "center",
        "paddingLeft": 30,
        "paddingRight": 30,
        "textAlign": "center",
      }
    }
  />
</View>
`;

exports[`when <Button /> is mounted renders green theme correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <BVLinearGradient
    colors={
      Array [
        4286430243,
        4290445056,
      ]
    }
    endPoint={
      Object {
        "x": 1,
        "y": 0,
      }
    }
    locations={null}
    startPoint={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    style={
      Object {
        "borderRadius": 35,
        "height": 55,
        "justifyContent": "center",
        "paddingLeft": 30,
        "paddingRight": 30,
        "textAlign": "center",
      }
    }
  />
</View>
`;

exports[`when <Button /> is mounted renders label correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#ffffff",
        "borderRadius": 35,
        "height": 55,
        "justifyContent": "center",
        "paddingLeft": 30,
        "paddingRight": 30,
        "textAlign": "center",
      }
    }
  >
    <Text
      adjustsFontSizeToFit={true}
      ellipsizeMode="tail"
      minimumFontScale={0.7}
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#000",
            "fontFamily": "Catamaran-Bold",
            "fontSize": 23,
            "letterSpacing": 0.5,
            "lineHeight": 39,
            "textAlign": "center",
            "textTransform": "uppercase",
          },
          undefined,
        ]
      }
    >
      My CTA
    </Text>
  </View>
</View>
`;

exports[`when <Button /> is mounted renders miniLightWhite theme correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Object {
        "borderColor": "#ffffff",
        "borderRadius": 20,
        "borderWidth": 1,
        "height": 30,
        "justifyContent": "center",
        "paddingLeft": 15,
        "paddingRight": 15,
        "textAlign": "center",
      }
    }
  />
</View>
`;

exports[`when <Button /> is mounted renders no label button correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#ffffff",
        "borderRadius": 35,
        "height": 55,
        "justifyContent": "center",
        "paddingLeft": 30,
        "paddingRight": 30,
        "textAlign": "center",
      }
    }
  />
</View>
`;

exports[`when <Button /> is mounted renders yellow theme correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <BVLinearGradient
    colors={
      Array [
        4294944768,
        4294764288,
      ]
    }
    endPoint={
      Object {
        "x": 1,
        "y": 0,
      }
    }
    locations={null}
    startPoint={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    style={
      Object {
        "borderRadius": 35,
        "height": 55,
        "justifyContent": "center",
        "paddingLeft": 30,
        "paddingRight": 30,
        "textAlign": "center",
      }
    }
  />
</View>
`;
